velox/exec/fuzzer/AsyncDataCacheFuzzer.cpp

```diff
@@ -0,0 +1,354 @@
+
+using namespace facebook::velox::cache;
+using namespace facebook::velox::dwio::common;
+
+namespace facebook::velox::exec::test {
+namespace {
+
+class AsyncDataCacheFuzzer {
+ public:
+  explicit AsyncDataCacheFuzzer(size_t seed);
+
+  void go();
+
+ private:
+  void seed(size_t seed) {
+    currentSeed_ = seed;
+    rng_.seed(currentSeed_);
+  }
+
+  void reSeed() {
+    seed(rng_());
+  }
+
+  folly::IOThreadPoolExecutor* executor() {
+    static std::mutex mutex;
+    std::lock_guard<std::mutex> l(mutex);
+    if (!executor_) {
+      // We have up to 20 threads and 16 threads are used for reading so
+      // there are some threads left over for SSD background write.
+      executor_ = std::make_unique<folly::IOThreadPoolExecutor>(20);
+    }
+    return executor_.get();
+  }
+
+  void initializeDataFiles();
+
+  void initializeCache();
+
+  void initializeInputs();
+
+  void readInParallel();
+
+  void reset();
+
+  uint64_t read(uint32_t fileIdx, uint64_t offset, uint64_t length);
+
+  FuzzerGenerator rng_;
+  size_t currentSeed_{0};
+  std::shared_ptr<memory::MemoryPool> rootPool_{
+      memory::memoryManager()->addRootPool()};
+  std::shared_ptr<memory::MemoryPool> pool_{rootPool_->addLeafChild("leaf")};
+  std::vector<StringIdLease> fileIds_;
+  std::vector<size_t> fileSizes_;
+  std::vector<StringIdLease> groupIds_;
+  std::vector<std::shared_ptr<LocalReadFile>> readFiles_;
+  std::vector<std::unique_ptr<CachedBufferedInput>> inputs_;
+  std::shared_ptr<exec::test::TempDirectoryPath> tempDirectory_;
+  std::unique_ptr<memory::MemoryManager> manager_;
+  memory::MemoryAllocator* allocator_{};
+  std::unique_ptr<folly::IOThreadPoolExecutor> executor_;
+  std::shared_ptr<AsyncDataCache> cache_;
+};
+
+template <typename T>
+bool isDone(size_t i, T startTime) {
+  if (FLAGS_duration_sec > 0) {
+    const std::chrono::duration<double> elapsed =
+        std::chrono::system_clock::now() - startTime;
+    return elapsed.count() >= FLAGS_duration_sec;
+  }
+  return i >= FLAGS_steps;
+}
+
+// Calls func on 'numThreads' in parallel.
+template <typename Func>
+void runThreads(int32_t numThreads, Func func) {
+  std::vector<std::thread> threads;
+  threads.reserve(numThreads);
+  for (int32_t i = 0; i < numThreads; ++i) {
+    threads.push_back(std::thread([i, func]() { func(i); }));
+  }
+  for (auto& thread : threads) {
+    thread.join();
+  }
+}
+
+AsyncDataCacheFuzzer::AsyncDataCacheFuzzer(size_t initialSeed) {
+  seed(initialSeed);
+}
+
+void AsyncDataCacheFuzzer::initializeDataFiles() {
+  tempDirectory_ = exec::test::TempDirectoryPath::create();
+
+  // Create files with random sizes.
+  if (fileIds_.empty()) {
+    for (auto i = 0; i < FLAGS_num_files; ++i) {
+      auto fileName = fmt::format("{}/file_{}", tempDirectory_->getPath(), i);
+
+      const size_t fileSize = boost::random::uniform_int_distribution<uint64_t>(
+          FLAGS_min_file_bytes, FLAGS_max_file_bytes)(rng_);
+      auto writeFile = LocalWriteFile(fileName);
```

Comment:
use velox file system to open and write file? Thanks!

Reply:
This is already the velox file system, no?